Mathlib/AlgebraicGeometry/PrimeSpectrum/Basic.lean

```diff
@@ -534,6 +535,86 @@ def localizationMapOfSpecializes {x y : PrimeSpectrum R} (h : x ⤳ y) :
       exact (IsLocalization.map_units (Localization.AtPrime x.asIdeal)
         ⟨a, show a ∈ x.asIdeal.primeCompl from h ha⟩ : _))
 
+section stableUnderSpecialization
+
+variable {R S : Type*} [CommRing R] [CommRing S] (f : R →+* S)
+
+lemma isClosed_range_of_stableUnderSpecialization
+    (hf : StableUnderSpecialization (Set.range (comap f))) :
+    IsClosed (Set.range (comap f)) := by
+  refine (isClosed_iff_zeroLocus _).mpr ⟨RingHom.ker f, le_antisymm ?_ ?_⟩
+  · rintro _ ⟨q, rfl⟩
+    exact Ideal.comap_mono bot_le
+  · intro p hp
+    obtain ⟨q, hq, hqle⟩ := Ideal.exists_minimalPrimes_le hp
+    obtain ⟨q', hq', hq'c⟩ := Ideal.exists_minimalPrimes_comap_eq f q hq
+    exact hf ((le_iff_specializes ⟨q, hq.1.1⟩ p).mp hqle) ⟨⟨q', hq'.1.1⟩, PrimeSpectrum.ext hq'c⟩
+
+@[stacks 05JL]
+lemma isClosed_image_of_stableUnderSpecialization
+    (Z : Set (PrimeSpectrum S)) (hZ : IsClosed Z)
+    (hf : StableUnderSpecialization (comap f '' Z)) :
+    IsClosed (comap f '' Z) := by
+  obtain ⟨I, rfl⟩ := (PrimeSpectrum.isClosed_iff_zeroLocus_ideal Z).mp hZ
+  have : (comap f '' zeroLocus I) = Set.range (comap ((Ideal.Quotient.mk I).comp f)) := by
+    rw [comap_comp, ContinuousMap.coe_comp, Set.range_comp, range_comap_of_surjective, Ideal.mk_ker]
+    exact Ideal.Quotient.mk_surjective
+  rw [this] at hf ⊢
+  exact isClosed_range_of_stableUnderSpecialization _ hf
+
+variable {f} in
+@[stacks 05JL]
+lemma stableUnderSpecialization_range_iff :
+    StableUnderSpecialization (Set.range (comap f)) ↔ IsClosed (Set.range (comap f)) :=
```

Comment:
These are closer to [00HY](https://stacks.math.columbia.edu/tag/00HY) which is about prime spectra, while [05JL](https://stacks.math.columbia.edu/tag/05JL) is about schemes. I'm changing the labels in #20533.